README.md
# PyCPS

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Downloads](https://static.pepy.tech/badge/pycpsdata)](https://pepy.tech/project/pycpsdata)

## Overview

Python package for loading [Current Population Survey (CPS)](https://www.census.gov/programs-surveys/cps/about.html) microdata into a [pandas DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) using the Census Bureau Data API, including [basic monthly CPS](https://www.census.gov/data/datasets/time-series/demo/cps/cps-basic.html) and [CPS ASEC](https://www.census.gov/data/datasets/time-series/demo/cps/cps-asec.html) microdata.

Note: This product uses the Census Bureau Data API but is not endorsed or certified by the Census Bureau.

For an R version of this package, check out [cpsR](https://github.com/matt-saenz/cpsR).

## Setup Instructions

Install the package:

```shell
pip install pycpsdata  # Alas, pycps was taken
```

and store your [Census API key](https://api.census.gov/data/key_signup.html) in an environment variable named `CENSUS_API_KEY`.

## Example Usage

```python
from pycps import get_asec

asec = get_asec(2021, ["a_age", "marsupwt"])

asec
#         a_age  marsupwt
# 0          56    687.71
# 1          57    687.71
# 2          78    646.86
# 3          65   1516.95
# 4          66   1516.95
# ...       ...       ...
# 163538     69    514.11
# 163539     70    516.25
# 163540     66    516.25
# 163541     55    386.37
# 163542     52    386.37
#
# [163543 rows x 2 columns]

asec.marsupwt.sum()
# 326195439.67
```


NEWS.md
# PyCPS 1.0.1

* Columns that are not parsable as numeric are now returned as strings. Previously, if any columns failed to parse as numeric an exception was raised. (https://github.com/matt-saenz/PyCPS/pull/3)

# PyCPS 1.0.0

* `get_asec()` and `get_basic()` no longer set an upper limit on supported years.

# PyCPS 0.2.0

* `get_asec()` now supports CPS ASEC microdata for 1992 to 2013.
* `get_basic()` now supports basic monthly CPS microdata for 2023.

# PyCPS 0.1.0

* Initial release.


lint_and_test.sh
#!/usr/bin/env bash
set -e

black --check --diff src tests
isort --check --diff --profile black src tests
mypy --strict src tests
flake8 --extend-ignore E501 src tests
pip install --quiet .
pytest --quiet tests


tests/test_get_data.py
"""Testing for the pycps.get_data module."""


from typing import Any

import pandas as pd  # type: ignore
import pytest

from pycps import get_data


class TestCheckVariables:
    def test_pass(self) -> None:
        get_data._check_variables(["column", "COLUMN_2", "3_last_column"])

    @pytest.mark.parametrize(
        "bad_type",
        [
            "column",
            ["column", 1],
        ],
    )
    def test_bad_type(self, bad_type: Any) -> None:
        with pytest.raises(TypeError):
            get_data._check_variables(bad_type)

    @pytest.mark.parametrize(
        "bad_value",
        [
            ["a column"],
            ["column?"],
            ["column", "column"],
            ["column", "COLUMN"],
            ["some_column,another_column"],
        ],
    )
    def test_bad_value(self, bad_value: list[str]) -> None:
        with pytest.raises(ValueError):
            get_data._check_variables(bad_value)


def test_make_url() -> None:
    actual = get_data._make_url(
        dataset="basic",
        year=2022,
        month=3,
        variables=["prtage", "pwsswgt"],
        key="1234",
    )

    expected = (
        # This technically works, although package users must set API key
        "https://api.census.gov/data/2022/cps/basic/mar?get=PRTAGE,PWSSWGT"
        # Fake key, breaks the above
        "&key=1234"
    )

    assert actual == expected


@pytest.mark.parametrize(
    "raw_data,expected_df",
    [
        (
            # All columns parsable as numeric
            [
                ["SOME_COLUMN", "ANOTHER_COLUMN"],
                ["9", "7.3"],
                ["5", "1.4"],
            ],
            pd.DataFrame(
                {
                    "some_column": [9, 5],
                    "another_column": [7.3, 1.4],
                }
            ),
        ),
        (
            # One column not parsable as numeric
            # Should remain a string
            [
                ["SOME_COLUMN", "ANOTHER_COLUMN", "STRING_COLUMN"],
                ["9", "7.3", "a"],
                ["5", "1.4", "b"],
            ],
            pd.DataFrame(
                {
                    "some_column": [9, 5],
                    "another_column": [7.3, 1.4],
                    "string_column": ["a", "b"],
                }
            ),
        ),
    ],
)
def test_build_df(
    raw_data: list[list[str]],
    expected_df: pd.DataFrame,
) -> None:
    actual_df = get_data._build_df(raw_data)

    pd.testing.assert_frame_equal(actual_df, expected_df)


class TestCheckYear:
    @pytest.mark.parametrize("year", [2000, 2001])
    def test_pass(self, year: int) -> None:
        get_data._check_year(year, min_year=2000)

    def test_fail(self) -> None:
        with pytest.raises(ValueError):
            get_data._check_year(1999, min_year=2000)


class TestCheckMonth:
    @pytest.mark.parametrize(
        "valid_month",
        range(1, 13),
    )
    def test_pass(self, valid_month: int) -> None:
        get_data._check_month(valid_month)

    @pytest.mark.parametrize(
        "bad_value",
        [0, 13],
    )
    def test_bad_value(self, bad_value: int) -> None:
        with pytest.raises(ValueError):
            get_data._check_month(bad_value)

    def test_bad_type(self) -> None:
        with pytest.raises(TypeError):
            get_data._check_month("jan")  # type: ignore


pycps.py
import os
import re

import pandas as pd  # type: ignore
import requests


BASE_URL = "https://api.census.gov/data/"


# Core functions


def get_asec(year: int, vars: list[str], show_url: bool = False) -> pd.DataFrame:
    """Get CPS ASEC microdata using the Census API."""

    key = _get_key()

    _check_year(year, dataset="asec")

    formatted_vars = _format_vars(vars)

    url = f"{BASE_URL}{year}/cps/asec/mar?get={formatted_vars}&key={key}"

    print(f"Getting CPS ASEC microdata for {year}")

    df = _get_data(url, show_url)

    return df


def get_basic(
    year: int, month: int, vars: list[str], show_url: bool = False
) -> pd.DataFrame:
    """Get basic monthly CPS microdata using the Census API."""

    key = _get_key()

    _check_year(year, dataset="basic")

    month_name, month_abb = _get_month_info(month)

    formatted_vars = _format_vars(vars)

    url = f"{BASE_URL}{year}/cps/basic/{month_abb}?get={formatted_vars}&key={key}"

    print(f"Getting basic monthly CPS microdata for {month_name} {year}")

    df = _get_data(url, show_url)

    return df


# Helpers


class CensusAPIRequestError(Exception):
    """Raise if Census API request fails."""


def _get_data(url: str, show_url: bool) -> pd.DataFrame:
    if show_url:
        # Suppress key!
        print("URL:", re.sub("&key=.*", "", url))

    resp = requests.get(url)

    if resp.status_code != 200:
        raise CensusAPIRequestError(
            f"Census API request failed [{resp.status_code}]: {resp.reason}"
        )

    if resp.headers["content-type"] != "application/json;charset=utf-8":
        raise CensusAPIRequestError("Census API did not return JSON")

    df = pd.DataFrame(resp.json())

    df.columns = df.iloc[0].str.lower().to_list()
    df = df.iloc[1:].reset_index(drop=True)
    df = df.apply(pd.to_numeric)

    return df


def _check_year(year: int, dataset: str) -> None:
    years_lookup = {"asec": (2014, 2021), "basic": (1994, 2022)}

    start_year, end_year = years_lookup[dataset]

    if year not in range(start_year, end_year + 1):
        raise ValueError(f"Years {start_year} to {end_year} are currently supported")


def _format_vars(vars: list[str]) -> str:
    if not isinstance(vars, list):
        raise TypeError("vars must be a list")

    not_string = [not isinstance(var, str) for var in vars]

    if any(not_string):
        raise TypeError("vars must only contain strings")

    invalid_char = [bool(re.search("[^A-Za-z0-9_]", var)) for var in vars]

    if any(invalid_char):
        raise ValueError(
            "Elements of vars must only contain letters, digits, and underscores"
        )

    if len(vars) != len(set(vars)):
        raise ValueError("vars must not contain any duplicate elements")

    formatted_vars = ",".join(vars).upper()

    return formatted_vars


# Create custom exception since clear built-in does not exist
class EnvVarNotFoundError(Exception):
    """Raise if environment variable is not found."""


def _get_key() -> str:
    key = os.getenv("CENSUS_API_KEY")

    if key is None:
        raise EnvVarNotFoundError(
            "You must provide a Census API key by setting env var CENSUS_API_KEY"
        )

    return key


def _get_month_info(month: int) -> tuple[str, str]:
    if month not in range(1, 13):
        raise ValueError("month must be a number ranging from 1 to 12")

    month_names = [
        "",  # Empty string at 0 so that month number matches month name
        "January",
        "February",
        "March",
        "April",
        "May",
        "June",
        "July",
        "August",
        "September",
        "October",
        "November",
        "December",
    ]

    month_name = month_names[month]
    month_abb = month_name[:3].lower()

    return month_name, month_abb


if __name__ == "__main__":
    # Get inputs
    print(
        "Hello! This if-name-main code calculates the employment-to-population",
        "(EPOP) ratio for a given month and year.",
    )

    month_year = input(
        "Please provide a month and year in MM/YYYY format (e.g., 09/2021): "
    )

    month, year = [int(x) for x in month_year.split("/")]
    month_name, month_abb = _get_month_info(month)

    # Get data
    print()  # For empty line
    cps = get_basic(year, month, ["prpertyp", "prtage", "pemlr", "pwcmpwgt"], True)
    print("\nRaw data:", cps, sep="\n")

    # Clean data
    cps = cps.loc[(cps.prpertyp == 2) & (cps.prtage >= 16)]
    cps["pop16plus"] = True  # Given above filter
    cps["employed"] = cps.pemlr.isin([1, 2])

    # Analyze data
    results = (
        cps[["pop16plus", "employed"]]
        .apply(lambda x, wt: x.dot(wt), wt=cps.pwcmpwgt)  # Weighted sum
        .astype(int)
    )

    print("\nWeighted sums:", results, sep="\n")

    # Calculate EPOP ratio
    print(
        f"\nThe EPOP ratio for {month_name} {year} was",
        f"{results['employed'] / results['pop16plus']:.1%}.",
    )


src/pycps/get_data.py
"""
Functions for loading CPS microdata into pandas DataFrames using the
Census API.
"""


# Key references:
# - General workings of the Census API:
#   https://www.census.gov/data/developers/guidance/microdata-api-user-guide.html
# - Easy way to check available years:
#   https://data.census.gov/mdat/#/


import os
import re
from http import HTTPStatus

import pandas as pd  # type: ignore
import requests  # type: ignore

# Hard code month names since they should not vary by locale
MONTH_NAME = [
    "",  # Empty string at 0 so that month number matches month name
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
]


def get_asec(
    year: int,
    variables: list[str],
) -> pd.DataFrame:
    """
    Load CPS ASEC microdata into a pandas DataFrame using the
    Census API.

    year: Year of data to retrieve. Years 1992 and on are currently
        supported.
    variables: List of variables to retrieve.
    """

    key = _get_key()  # Get key first to fail fast if it is not found

    _check_year(year, min_year=1992)

    month = 3  # Month of CPS ASEC is always March

    url = _make_url("asec", year, month, variables, key)

    df = _get_data(url)

    return df


def get_basic(
    year: int,
    month: int,
    variables: list[str],
) -> pd.DataFrame:
    """
    Load basic monthly CPS microdata into a pandas DataFrame using the
    Census API.

    year: Year of data to retrieve. Years 1989 and on are currently
        supported.
    month: Month of data to retrieve (specified as a number).
    variables: List of variables to retrieve.
    """

    key = _get_key()

    _check_year(year, min_year=1989)

    url = _make_url("basic", year, month, variables, key)

    df = _get_data(url)

    return df


def _make_url(
    dataset: str,
    year: int,
    month: int,
    variables: list[str],
    key: str,
) -> str:
    """Create URL for Census API request."""

    _check_month(month)
    _check_variables(variables)

    month_abb = MONTH_NAME[month][:3].lower()
    vars_string = ",".join(variables).upper()

    url = (
        "https://api.census.gov/data/"
        f"{year}/cps/{dataset}/{month_abb}?get={vars_string}&key={key}"
    )

    return url


def _get_data(url: str) -> pd.DataFrame:
    """Send request to URL and return response content as DataFrame."""

    headers = {"user-agent": "https://github.com/matt-saenz/PyCPS"}

    resp = requests.get(url, headers=headers)

    if resp.status_code != 200:
        # Census API does not (at least currently) supply resp reason
        resp_reason = HTTPStatus(resp.status_code).phrase

        raise CensusAPIRequestError(
            f"Census API request failed [{resp.status_code}]: {resp_reason}"
        )

    if resp.headers["content-type"] != "application/json;charset=utf-8":
        raise CensusAPIRequestError("Census API did not return JSON")

    raw_data = resp.json()

    if not isinstance(raw_data, list):
        raise CensusAPIRequestError("Census API data not parsed as expected")

    df = _build_df(raw_data)

    return df


def _build_df(raw_data: list[list[str]]) -> pd.DataFrame:
    """Build DataFrame out of parsed response content."""

    column_names = [column_name.lower() for column_name in raw_data[0]]
    rows = raw_data[1:]

    df = pd.DataFrame(data=rows, columns=column_names)

    # Set errors to "ignore" so that if column fails to parse as
    # numeric, it will remain a string
    # Originally flagged in:
    # https://github.com/matt-saenz/PyCPS/pull/3
    df = df.apply(pd.to_numeric, errors="ignore")

    return df


def _check_year(year: int, min_year: int) -> None:
    """Check if year is valid."""

    if year < min_year:
        raise ValueError(f"Years {min_year} and on are currently supported")


def _check_variables(variables: list[str]) -> None:
    """Check if variables list is valid."""

    if not isinstance(variables, list):
        raise TypeError("variables must be a list")

    for var in variables:
        if not isinstance(var, str):
            raise TypeError("variables must only contain strings")

    for var in variables:
        if re.search(r"[^A-Za-z0-9_]", var):
            raise ValueError(
                "Elements of variables must only contain letters, digits, and underscores"
            )

    # Ensure all lowercase for dup checking
    variables = [var.lower() for var in variables]

    if len(variables) != len(set(variables)):
        raise ValueError("variables must not contain any duplicate elements")


def _check_month(month: int) -> None:
    """Check if month is valid."""

    if not isinstance(month, int):
        raise TypeError("month must be an int")

    if month not in range(1, 13):
        raise ValueError("month must range from 1 to 12")


def _get_key() -> str:
    """Get user's Census API key."""

    key = os.getenv("CENSUS_API_KEY")

    if key is None:
        raise EnvVarNotFoundError(
            "You must provide a Census API key by setting env var CENSUS_API_KEY"
        )

    return key


# Custom exceptions


class CensusAPIRequestError(Exception):
    """Raise if something goes wrong with a Census API request."""


class EnvVarNotFoundError(Exception):
    """Raise if an environment variable is not found."""
